[TBDGen] Fix symbol mismatch for enum case constructors

theblixguy opened the pull request:
For the following code:

```swift
protocol P {
  static var bar1: Self { get }
  static func bar2(arg: Int) -> Self
}

enum E: P {
  case bar1
  case bar2(arg: Int)
}
```

TBDGen validation fails with an error: `symbol '<case constructor symbol here>' is in generated IR file, but not in TBD file` when we pass the `-enable-testing` flag. It seems like it's looking for the enum case constructor symbols, which are not present when testing is enabled.

Resolves SR-12821
Resolves rdar://problem/63364542

slavapestov commented on lib/TBDGen/TBDGen.cpp:
```diff
@@ -496,10 +496,15 @@ void TBDGenVisitor::addConformances(DeclContext *DC) {
+      // exported.
+      if (isa<EnumElementDecl>(witnessDecl) &&
+          SwiftModule->isTestingEnabled()) {
+        addSymbol(SILDeclRef(witnessDecl, SILDeclRef::Kind::EnumElement));
```
I don't think this is the best fix. The case constructors should not be public with -enable-testing. I think they need to be emitted with shared linkage, not hidden, in the case that the enum is `internal` at the AST level. Can you take a look at SILDeclRef::getLinkage() and see if you can fix it there?

theblixguy replied:
Thank you for the suggestion! Adding a check to `getLinkage` has no effect unfortunately:

```c++
if (isEnumElement() && d->getFormalAccess() == AccessLevel::Internal) {
  return SILLinkage::Shared;
}
```

theblixguy replied:
Sorry, I forgot I made the protocol and enum `public`. When I make them `internal`, I get a crash instead in `addConformances` because its calling `mangleWitnessThunk` which expects the requirement to be a function or constructor. Removing the call to `addSymbolIfNecessary` fixes the problem but then you get the same missing symbol error for `$s4test1EOAA1PA2aDP4bar1xvgZTW` which is the witness thunk.

slavapestov replied:
Enum elements need to be handled the same as on-demand accessors. See this block:

```
  // Forced-static-dispatch functions are created on-demand and have
  // at best shared linkage.
  if (auto fn = dyn_cast<FuncDecl>(d)) {
    if (fn->hasForcedStaticDispatch()) {
      limit = Limit::OnDemand;
    }
  }
```

You just need to add another check, like

```
if (isEnumElement())
   limit = Limit::OnDemand;
```

theblixguy replied:
Hmm,  I added this exact check before (commit https://github.com/apple/swift/pull/28916/commits/7d3cf178f125c07dd3908cfb86cb884d94c0cea2) but I removed it for some reason 🤷

slavapestov replied:
Ah, nice catch! Yeah, let's bring this back, and cherry-pick the fix back to 5.3 too.

theblixguy replied:
Okay, I have added those checks. But, it seems like the combination of the checks is preventing `TBDGenVisitor::addConformances` from adding the symbol for the witness of the property requirement (which is an enum case without payload). I can't use `addSymbolIfNecessary` anymore if the requirement is a property and adding a symbol by passing the mangled name doesn't seem to work as well.

slavapestov commented:
This example directly demonstrates the problem with making these symbols public:

```
// a.swift
internal enum E {
  case a
}

// b.swift
protocol P1 {
  static var a: Self { get }
}

extension E : P1 {}

// c.swift
protocol P2 {
  static var a: Self { get }
}

extension E : P2 {}
```

Since the case constructors are emitted on demand, they must either be private in SIL, or shared. Compiling the above fails with:

```
duplicate symbol '_$s4main1EO1ayA2CmF' in:
    /var/folders/wg/nmvf884n4b14lt0yw4khjsv00000gn/T/qq1-e8f1a0.o
    /var/folders/wg/nmvf884n4b14lt0yw4khjsv00000gn/T/qq2-28e0f5.o
```

slavapestov commented:
This is a regression. I had removed enum element constructor thunks because they were no longer used when emitting a partial application of an enum case. However, I added them back since they are now used as protocol witnesses, but I forgot to bring back the change to SILDeclRef::getLinkage().

slavapestov commented:
In the future though, it would be cool if Sema's witness matching would build the AST for the witness call (it already uses a constraint solver to match the types so this is not a big step). Then SILGen can emit the Expr when it emits the witness thunk. This would eliminate this special-purpose enum constructor thunk code path, as well as dynamic thunks which are also used for a similar purpose.

theblixguy commented:
> This is a regression. I had removed enum element constructor thunks because they were no longer used when emitting a partial application of an enum case. However, I added them back since they are now used as protocol witnesses, but I forgot to bring back the change to SILDeclRef::getLinkage().

No worries! Now we have a test case to prevent future regressions :) 

> In the future though, it would be cool if Sema's witness matching would build the AST for the witness call (it already uses a constraint solver to match the types so this is not a big step).

Great! I can give it a shot, although I would need to understand your idea better - do you mean we should effectively build the thunk in the AST and call that?

slavapestov commented on lib/SIL/IR/SILDeclRef.cpp:
```diff
@@ -256,6 +256,12 @@ SILLinkage SILDeclRef::getLinkage(ForDefinition_t forDefinition) const {
   if (isClangImported())
     return SILLinkage::Shared;
 
+  // Enum case constructors have shared linkage if the enum is internal.
```
This part should not be necessary, just setting the limit for all enum elements is sufficient.

theblixguy replied:
Sorry, I thought we needed both. I have removed it now!

slavapestov commented:
> Okay, I have added those checks. But, it seems like the combination of the checks is preventing TBDGenVisitor::addConformances from adding the symbol for the witness of the property requirement (which is an enum case without payload).

Do you mean the symbol for the witness thunk? The witness itself here is the enum constructor thunk, and no combination of build flags should ever make it public -- it's `shared [serialized]` which means it's part of the SIL module and emitted into any client that uses it.

The witness thunk on the other hand will be public if the conformance is `internal` at the AST level, and you build with `-enable-testing`, but the logic for that should already be in place and this behavior should not depend on whether the witness is an enum case or not.

Also when the protocol requirement is a stored property (static or otherwise), the witness thunks correspond to the _accessors_, not the property itself.

This is all quite confusing and complicated, I know :)

theblixguy commented:
Thanks for the explanation! I guess so:

> symbol '$s4test1EOAA1PA2aDP4bar1xvgZTW' (protocol witness for static test.P.bar1.getter : A in conformance test.E : test.P in test) is in generated IR file, but not in TBD file

Initially, the missing symbols were for the case constructor (if you look at the original bug report) I think (according to the generated SIL code). But now with this fix, calling `addSymbolIfNecessary` triggers the assert in it (`(isa<FuncDecl>(Requirement) && "expected function")`).

So, I think I need to figure out a different way to add these symbols (or maybe there's a bug elsewhere).

slavapestov commented:
`mangleWitnessThunk()` expects that the requirement is either a method or a constructor. For a property or subscript requirement, we emit witness thunks for each accessor. There's no witness thunk for the property or subscript requirement itself.

Since an enum element can only witness a read-only property requirement, you need to call `addSymbolIfNecessary()` with the getter as the requirement, and the enum element as the witness.

theblixguy commented:
That makes sense! Okay, I have tweaked the logic to pass the getter as requirement instead. We still have some missing symbols, but instead of missing from the TBD file, they're missing from the IR file. Is that a problem?

> symbol '_$s4test1EO4bar1yA2CmFWC' (enum case for test.E.bar1(test.E.Type) -> test.E) is in TBD file, but not in generated IR

> symbol '_$s4test1EO4bar2yACSi_tcACmFWC' (enum case for test.E.bar2(test.E.Type) -> (Swift.Int) -> test.E) is in TBD file, but not in generated IR

slavapestov commented:
Yes, that is a problem. The set of symbols generated by IRGen should match TBDGen exactly. Do they disagree on the whether the symbol should be public for some reason?

slavapestov commented:
Does the TBDGen problem occur if the enum and protocol are both public and you build without -enable-testing as well?

theblixguy commented:
Yeah, it does.

theblixguy commented:
I was just looking at the code you removed in [this](https://github.com/apple/swift/pull/30499/commits/9ec80df97ed174e4cf85b036239e5980df3606ca) commit from `getLinkage` and the check I have added is effectively the same I think, so I think something else regressed elsewhere since SE-0280 landed.

slavapestov commented:
Ah, I see. This is a bug in TBDGen. Look at `visitEnumElementDecl()`:

```
addSymbol(LinkEntity::forEnumCase(EED));
```

We should only add this symbol if `EED->getEnum()->isResilient()`.

Also there's a dead `isResilient()` check in `visitEnumDecl()` you may as well remove.

theblixguy commented:
Ah, that makes sense, we were adding the symbol in TBDGen regardless of whether the parent enum is resilient or not. Thanks! I was also (wrongly) looking at linkage checks in TBDGen rather than resilience. The dead code in `visitEnumDecl` was a nice hint.

slavapestov commented:
> Edit: Okay, seems like we need to use a SILDeclRef when calling addSymbol instead of a LinkEntity.

No, those are two different symbols. The SILDeclRef mangles as the thunk that is called to construct the enum value. The LinkEntity is a global variable storing the tag. The former symbol should never be public, and thus does not appear in the TBD file. The latter is only emitted if the enum is resilient.

theblixguy commented:
Oh, okay! I didn't know that. Thanks for the explanation! It seems like I didn't had to make that change at all, something was wrong with my build. I think the previous change was enough as the tests now pass locally for me. Let me revert that commit.

theblixguy committed:
[SIL] Use Limit::OnDemand for an enum element in SILDeclRef::getLinkage

theblixguy committed:
[TBDGen] Emit case symbols only if enum is resilient and some other fixes

(1) Pass property getter as requirement when the witness is an payload-less enum case (2) Remove dead code from visitEnumDecl

theblixguy committed:
[Test] Add some TBDGen test cases for enum case witnesses

slavapestov commented:
Thanks. The change looks great! Do you mind also cherry-picking it to the 5.3 branch?

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

theblixguy committed:
[Test] Update a SILGen test

theblixguy commented:
@swift-ci please smoke test

theblixguy commented:
@swift-ci Please test source compatibility

govom2p commented:
I do not understand anything about github and language is my obstacle tell me what github is for

Được gửi từ iPhone của tôi

> Vào ngày 19 thg 5, 2020, lúc 20:09, Suyash Srijan <notifications@github.com> đã viết:
> 
> ﻿
> @swift-ci please smoke test
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
